### PR TITLE
[stdlib] Fix the `String.decodeCString` for UTF16 and UTF32

### PR DESCRIPTION
**Pull request by moiseev**
#### What's in this pull request?

Essentially the problem was that `strlen` is not the right way of
obtaining a length of anything but null-terminated UTF-8 sequence of
characters. Other encodings require alternative mechanisms.
#### Resolved bug number: ([SR-1578](https://bugs.swift.org/browse/SR-1578))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@gribozavr, @dabrahams please review.

**Comment by moiseev** [comment]
@swift-ci Please test Linux platform

File: stdlib/public/core/Unicode.swift [hunk 1]
```diff
@@ -1149,6 +1160,24 @@ extension UnicodeScalar {
   }
 }
 
+extension UnicodeCodec where CodeUnit : UnsignedInteger {
+  public static func _nullCodeUnitOffset(_ input: UnsafePointer<CodeUnit>) -> Int {
+    var input = input
+    var length = 0
+    while input.pointee != 0 {
```
Comment (gribozavr) [inline stdlib/public/core/Unicode.swift]: `input[length] != 0` and then there's no need to mutate `input`.

**Comment by gribozavr** [comment]
LGTM.

**Comment by gribozavr** [comment]
Updates LGTM as well.

**Pull request merged**